cpp/src/digitize/digitize.cu

```diff
@@ -0,0 +1,152 @@
+  auto operator()(column_view const& col, column_view const& bins, bool upper_bound,
+                  rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    // TODO allocate null mask and copy contents from col?
+    auto output = cudf::make_numeric_column(data_type{INT32}, col.size(),
+      cudf::UNALLOCATED, stream, mr);
+    auto output_view = output->mutable_view();
+
+    // Allocate index columns to be used with element_relation_comparator
+    auto col_indices = make_sequence(col.size(), mr, stream);
+    auto col_indices_view = col_indices->view();
+    auto bins_indices = make_sequence(bins.size(), mr, stream);
+    auto bins_indices_view = bins_indices->view();
+
+    auto col_device_view = column_device_view::create(col, stream);
+    auto bins_device_view = column_device_view::create(bins, stream);
+    auto comparator = less_equal_wrapper<true>(
+        *bins_device_view, *col_device_view, _null_precedence);
+
+    if (upper_bound) {
+      thrust::upper_bound(rmm::exec_policy(stream)->on(stream),
+        bins_indices_view.begin<int32_t>(), bins_indices_view.end<int32_t>(),
+        col_indices_view.begin<int32_t>(), col_indices_view.end<int32_t>(),
+        output_view.begin<int32_t>(), comparator);
+    } else {
+      thrust::lower_bound(rmm::exec_policy(stream)->on(stream),
+        bins_indices_view.begin<int32_t>(), bins_indices_view.end<int32_t>(),
+        col_indices_view.begin<int32_t>(), col_indices_view.end<int32_t>(),
+        output_view.begin<int32_t>(), comparator);
+    }
+
+    return output;
+  }
+};
+
+struct binary_search_bound {
+  template<typename T>
+  auto operator()(column_view const& col, column_view const& bins, bool upper_bound,
+                  rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    auto output = cudf::make_numeric_column(data_type{INT32}, col.size(),
+      cudf::UNALLOCATED, stream, mr);
+    auto output_view = output->mutable_view();
+
+    if (upper_bound) {
+      thrust::upper_bound(rmm::exec_policy(stream)->on(stream),
+        bins.begin<T>(), bins.end<T>(), col.begin<T>(), col.end<T>(),
+        output_view.begin<int32_t>(), thrust::less_equal<T>());
+    } else {
+      thrust::lower_bound(rmm::exec_policy(stream)->on(stream),
+        bins.begin<T>(), bins.end<T>(), col.begin<T>(), col.end<T>(),
+        output_view.begin<int32_t>(), thrust::less_equal<T>());
+    }
+
+    return output;
+  }
+};
+
+}  // namespace
+
+namespace detail {
+
+std::unique_ptr<column>
+digitize(column_view const& col, column_view const& bins, bool right,
+         rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  auto const dtype = col.type();
+  CUDF_EXPECTS(dtype == bins.type(), "Column type mismatch");
+
+  // TODO should this make use of the comparable type traits?
+  CUDF_EXPECTS(is_numeric(dtype) || is_timestamp(dtype), "Type must be numeric or timestamp");
+
+  CUDF_EXPECTS(0 == bins.null_count(), "Bins column must not have nulls");
```

Comment:
Does this need to be true?

Reply:
The bins column is expected to be in sorted order (it's used for binary search), and having nulls in the middle would violate the sorted order. Maybe I could use a sequence iterator that would skip indices of nulls in the bins column?

Reply:
But you can have null values in a total sorted order of values, they'd be either at the beginning or the end. As such, I don't think we need to enforce that `bins.null_count() == 0`

Reply:
Ok, removed the check and have it taking the have_nulls branch if either cols or bins have nulls.